### ISSUE TITLE
add JPMS metadata to gradle jars

### ISSUE THREAD

**Issue by xenoterracide** [gradle/gradle#31028]
### Expected Behavior

It'd be nice if gradle jars were at least JPMS automatic modules themselves. module-info would be preferred, but that is harder to perhaps even impossible at this time.

### Context

This is so that plugin authors could choose to `module-info` for compile time restrictions if they wanted. 

Having gradle eventually require all plugins to have a module-info would solve the duplicate library classpath issue (or at least make it obvious). I don't expect that to happen for a very long time.

**Comment by ljacomet** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

***

It is very unlikely that Gradle will itself execute with the module path concept. Also our current plans on publishing a Gradle API are about publishing a single JAR for this.
There would be limited value in pretending it can be consumed in a module aware runtime.

**Comment by xenoterracide** [comment]
Yeah, really right now I would just like to see the automatic module metadata added. That way if we want to use more statically typed module metadata we can at compile time.

**Comment by asodja** [comment]
We have no plan to expose Gradle API through JPMS at this stage. We will progressively remove internal APIs from public API jar. See also https://github.com/gradle/gradle/issues/29483

We will investigate JPMS for Gradle API in:
- #34403

**Comment by xenoterracide** [comment]
Oh well would have been nice to have had an automatic module name for API in the meantime so that it could have been used for people who want to use compile time restrictions for their plugins.
